### PR TITLE
WIP: Use default :enum value from database column's one

### LINKED ISSUE #23187: Setting enum default via ':default' on DB field is broken [bisected]
[Testcase](https://gist.github.com/ojab/6a5c0853fda1cc37f20b).
If column for enum is created with default value:
in rails master: Model instance is built with field == nil
in rails-before-bad-commit: Model is built with filed == default value.

```
991c7b8cd69d7ba5e221a19e5f386e3ba02eb9d is the first bad commit
commit e991c7b8cd69d7ba5e221a19e5f386e3ba02eb9d
Author: Santiago Pastorino <santiago@wyeworks.com>
Date:   Thu Dec 31 19:45:47 2015 -0300

    value has the right type here

    using `enum kind: [ :any, :income, :expense ]` syntax, value is already an
    integer and using `enum kind: { any: 'any', income: 'income', expense:
    'expense' }` syntax value is a string.

    This allows us to define the mapping in the enum definition.

    This reverts commit 933decceaf6092020ba7ba768b51b2db9d5b882f.

:040000 040000 9d5c346b7de41f9b6cb0b9d23fd2ac5787fc0cbb 7a8f8fd4d6faaa43b779102fae172a874ce5b644 M	activerecord
bisect run success
```
ping @spastorino

### PR DESCRIPTION
**Pull request by ojab**
Reverts e991c7b8cd69d7ba5e221a19e5f386e3ba02eb9d and adds the test
Fixes #23187

### CODE REVIEW

**Comment by spastorino** [comment]
This is not the right fix.

Read https://github.com/rails/rails/commit/e991c7b8cd69d7ba5e221a19e5f386e3ba02eb9d

I will check this out.

**Comment by ojab** [comment]
hmm, basically sqlite (in this example) returns string zero as string for ['PRAGMA table_info(books)](https://github.com/rails/rails/blob/39ea2b09edcf8179fe69808424d325612194d4cb/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L478) with type integer and we're overwriting resulting `ActiveModel::Type::Integer` with `ActiveRecord::Enum::EnumType` in [load_schema!](https://github.com/rails/rails/blob/39ea2b09edcf8179fe69808424d325612194d4cb/activerecord/lib/active_record/attributes.rb#L229).

So basically we should add type_cast somewhere, but I'm not sure where exactly where.
Actually sqlite3 (and at least postgres, I suggest, because this is reproducible on postgres) returns all integer integers as string (for example `{"cid"=>1, "name"=>"author_id", "type"=>"integer", "notnull"=>0, "dflt_value"=>"0", "pk"=>0, 0=>1, 1=>"author_id", 2=>"integer", 3=>0, 4=>"0", 5=>0}`), but .attributes_before_type_cast returns integers:
```
(byebug) Book.create.attributes_before_type_cast
{"id"=>1, "author_id"=>0, "format"=>nil, "name"=>nil, "status"=>nil, "read_status"=>nil, "nullable_status"=>nil, "language"=>nil, "author_visibility"=>nil, "illustrator_visibility"=>nil, "font_size"=>nil}
```
and type_cast actually happens in [Attribute#value](https://github.com/rails/rails/blob/39ea2b09edcf8179fe69808424d325612194d4cb/activerecord/lib/active_record/attribute.rb#L36).

So the questions are:
- what exactly .attributes_before_type_cast should return?
- why type_cast happens when value is retrieval, not on value set?

**Comment by ojab** [comment]
So I think that basically type_cast should be moved to [define_default_attribute](https://github.com/rails/rails/blob/39ea2b09edcf8179fe69808424d325612194d4cb/activerecord/lib/active_record/attributes.rb#L243) or somewhere like that. I've pushed rails with this fix, please see the updated PR (count it as RFC, no commit message changes/etc was made). There is three test failures:
```
  1) Failure:
DefaultNumbersTest#test_default_positive_integer [/Users/ojab/Documents/repos/rails/activerecord/test/cases/defaults_test.rb:43]:
Expected: "7"
  Actual: 7


  2) Failure:
DefaultNumbersTest#test_default_decimal_number [/Users/ojab/Documents/repos/rails/activerecord/test/cases/defaults_test.rb:55]:
--- expected
+++ actual
@@ -1 +1 @@
-"2.78"
+#<BigDecimal:7fbe93366d58,'0.278E1',18(27)>



  3) Failure:
DefaultNumbersTest#test_default_negative_integer [/Users/ojab/Documents/repos/rails/activerecord/test/cases/defaults_test.rb:49]:
Expected: "-5"
  Actual: -5
```
but I'm not sure if this tests are correct.

If this fix looks proper -- I'll try to investigate if there is any more suitable place for type_cast and remove excessive type_casts further.

**Comment by ojab** [comment]
Aha, these tests are for before_type_cast values.
AFAICS there is inconsistency here and I don't know how rails should behave
```
(byebug) books(:tlg).dup.attributes_before_type_cast
{"id"=>nil, "author_id"=>1, "format"=>nil, "name"=>"Thoughtleadering", "status"=>"proposed", "read_status"=>"unread", "nullable_status"=>nil, "language"=>"english", "author_visibility"=>"visible", "illustrator_visibility"=>"visible", "font_size"=>"small"}
(byebug) books(:tlg).attributes_before_type_cast
{"id"=>4, "author_id"=>1, "format"=>nil, "name"=>"Thoughtleadering", "status"=>0, "read_status"=>0, "nullable_status"=>nil, "language"=>0, "author_visibility"=>0, "illustrator_visibility"=>0, "font_size"=>0}
```
Please advise, I'm lost.

File: activerecord/lib/active_record/attributes.rb [hunk 1]
```diff
@@ -251,7 +251,7 @@ def define_default_attribute(name, value, type, from_user:)
             _default_attributes[name],
           )
         else
-          default_attribute = Attribute.from_database(name, value, type)
+          default_attribute = Attribute.from_database(name, type.cast(value), type)
```
Comment (sgrif) [inline activerecord/lib/active_record/attributes.rb]: Attributes should never be cast on construction.

**Pull request closed without merging**